cdk/domino_cdk/provisioners/eks/eks_iam_roles_for_k8s.py

```diff
@@ -0,0 +1,170 @@
+            ),
+        )
+        statement_json['Condition']['StringLike'] = cdk.CfnJson(
+            self.scope, "OidcJson", value={f"{fn}:aud": "sts.amazonaws.com", f"{fn}:sub": "system:serviceaccount:*"}
+        )
+        managed_policies = {}
+        for name, cfg in ecr_policies.items():
+            managed_policies[name] = self.create_ecr_policy(stack_name, name, cfg)
+
+        managed_policies.update(self.create_s3_policies(stack_name, buckets))
+
+        for name, policy_list in roles.items():
+            iam_role = iam.Role(
+                self.scope,
+                f"{stack_name}-IAM-role-for-k8s-{name}",
+                # Undesired side effect of this: additional statement is created to trust this principal.
+                # Because Role constructor mandates principal and creates assume_role_policy
+                # with this statement. And later we cannot neither replace assume_role_policy nor remove this statement.
+                # Potentially we can avoid this by using CnfRole but no examples exist. A guesswork can take a week.
+                assumed_by=iam.ServicePrincipal('eks.amazonaws.com'),
+                role_name=f"{stack_name}-IAM-role-for-k8s-{name}",
+            )
+            iam_role.assume_role_policy.add_statements(iam.PolicyStatement.from_json(statement_json))
+            for policy_name in policy_list:
+                iam_role.add_managed_policy(managed_policies[policy_name])
+
+    def create_ecr_policy(self, stack_name: str, policy_name: str, actions: List[str]):
+        return iam.ManagedPolicy(
+            self.scope,
+            f"{stack_name}-{policy_name}-ECR",
+            managed_policy_name=f"{stack_name}-{policy_name}-ECR",
+            statements=[
+                iam.PolicyStatement(
+                    effect=iam.Effect.ALLOW,
+                    actions=actions,
+                    resources=["*"],
+                ),
+                iam.PolicyStatement(
+                    effect=iam.Effect.DENY,
+                    actions=["ecr:*"],
+                    conditions={"StringNotEqualsIfExists": {"ecr:ResourceTag/domino-deploy-id": stack_name}},
+                    resources=[f"arn:aws:ecr:*:{self.scope.account}:*"],
+                ),
+            ],
+        )
+
+    def create_s3_policies(self, stack_name: str, buckets: Dict[str, Bucket]) -> Dict[str, iam.ManagedPolicy]:
+        policies = {}
+        for name, bucket in buckets.items():
+            policies[f"read_{name}"] = iam.ManagedPolicy(
+                self.scope,
+                f"{stack_name}-read-{name}",
+                managed_policy_name=f"{stack_name}-read-{name}",
+                statements=[
+                    iam.PolicyStatement(
+                        actions=s3_read_permissions,
+                        resources=[f"{bucket.bucket_arn}*"],
+                    )
+                ],
+            )
+            policies[f"write_{name}"] = iam.ManagedPolicy(
```

Comment:
I'm not sure what alternatives I have to offer for the configuration paradigm, but I'd prefer not to rely on composed strings for purely internally accessed things if possible (ie as opposed to something like `policies["blobs"]["write"]`).

Reply:
I'm fine with this.

Reply:
Why it is done this way:

We want to have internal config with the roles that refer to the policies.

Best way to refer is by some reference ID, in this case, policy name.

If we put policies in some kind of language-based memory tree structure, we cannot refer them by the ID anymore.

Furthermore, since now I generate the bucket-based policies in runtime, based on the bucket list from the config, now I cannot have role definitions before the policies are programmatically generated.

An alternative to this is to predefine the whole policy tree structure before the provisioner is called. But then it is redundant and represents an anti-pattern of "second source of truth" for the bucket list. Then the bucket list will come both from the config and from internal roles config.

TBH I don't really understand your resistance to the concatenated strings. 
Even is simple hash map, if you have a key (a, b) , then internally if will be concatenated and become hash(a.concat(b)) anyway.
Creating of two level tree structure to just hold a list of the policies we want to refer in a human-readable way...

I could make a tree structure and refer the policies as some kind of 2-part formatted string like "blobs/read" and then parse the reference to split it into 2 parts to use as 2 indexes. But I think it is overkill to just to avoid the string concatenation.

Reply:
Actually, not a big change.

Would you agree on:
1. Put the policies in 2-level tree like s3_policies["blobs'']["read"]
2. Refer them as "blobs-read" in the flat policy list and then just to split("-")[0]

?

Reply:
Ah. I still need to concatenate the string to form the external policy name.

I.e. if we have stack denis-test and ["blobs"]["read"] we still create a policy "denis-test-blobs-read". I.e. stack_name + bucket_name + mode.

Would in invalidate your "for internal purposes only" objection?

Reply:
Ah. I still need to concatenate the string to form the external policy name.

I.e. if we have stack denis-test and ["blobs"]["read"] we still create a policy "denis-test-blobs-read". I.e. stack_name + bucket_name + mode.

Would in invalidate your "for internal purposes only" objection?

Reply:
I did say I'm not sure what alternatives I have to offer. :)

Something like the two level tree is what I was thinking, but still avoiding string manipulation (so do the same thing on the list side).

Here's a quick (completely untested) attempt: https://gist.github.com/Secretions/5d925ac0a876a7b4ee00b9ca49db5efe

(I might also reverse the order in the policy names if I did this approach, to match the structure, ie name-read, name-write)

Reply:
Yay-yay. Wait a min

Reply:
Undertested, but take a look

Reply:
Names:

```
┌───┬───────────────────────────────────────────────┬────────┬───────────────────────────────┬─────────────────────────────────────────────────┬────────────────────────────────────────────┐
│   │ Resource                                      │ Effect │ Action                        │ Principal                                       │ Condition                                  │
├───┼───────────────────────────────────────────────┼────────┼───────────────────────────────┼─────────────────────────────────────────────────┼────────────────────────────────────────────┤
│ + │ ${denis-test1-4SA-blobs-write.Arn}            │ Allow  │ sts:AssumeRole                │ Service:eks.amazonaws.com                       │                                            │
│ + │ ${denis-test1-4SA-blobs-write.Arn}            │ Allow  │ sts:AssumeRoleWithWebIdentity │ Federated:${eks/OpenIdConnectProvider/Resource} │ "StringLike": "${OidcJson/Resource.Value}" │
├───┼───────────────────────────────────────────────┼────────┼───────────────────────────────┼─────────────────────────────────────────────────┼────────────────────────────────────────────┤
│ + │ ${denis-test1-4SA-blobs-write--logs-read.Arn} │ Allow  │ sts:AssumeRole                │ Service:eks.amazonaws.com                       │                                            │
│ + │ ${denis-test1-4SA-blobs-write--logs-read.Arn} │ Allow  │ sts:AssumeRoleWithWebIdentity │ Federated:${eks/OpenIdConnectProvider/Resource} │ "StringLike": "${OidcJson/Resource.Value}" │
├───┼───────────────────────────────────────────────┼────────┼───────────────────────────────┼─────────────────────────────────────────────────┼────────────────────────────────────────────┤
│ + │ ${denis-test1-4SA-images-write.Arn}           │ Allow  │ sts:AssumeRole                │ Service:eks.amazonaws.com                       │                                            │
│ + │ ${denis-test1-4SA-images-write.Arn}           │ Allow  │ sts:AssumeRoleWithWebIdentity │ Federated:${eks/OpenIdConnectProvider/Resource} │ "StringLike": "${OidcJson/Resource.Value}" │
└───┴───────────────────────────────────────────────┴────────┴───────────────────────────────┴─────────────────────────────────────────────────┴────────────────────────────────────────────┘
IAM Policy Changes
┌───┬───────────────────────────────────────────┬───────────────────────────────┐
│   │ Resource                                  │ Managed Policy ARN            │
├───┼───────────────────────────────────────────┼───────────────────────────────┤
│ + │ ${denis-test1-4SA-blobs-write}            │ ${denis-test1-S3-blobs-write} │
├───┼───────────────────────────────────────────┼───────────────────────────────┤
│ + │ ${denis-test1-4SA-blobs-write--logs-read} │ ${denis-test1-S3-blobs-write} │
│ + │ ${denis-test1-4SA-blobs-write--logs-read} │ ${denis-test1-S3-logs-read}   │
├───┼───────────────────────────────────────────┼───────────────────────────────┤
│ + │ ${denis-test1-4SA-images-write}           │ ${denis-test1-ECR-ecr-write}  │
└───┴───────────────────────────────────────────┴───────────────────────────────┘
```

Reply:
What is your opinion on this particular difference between the gist and the commit you pushed?

```
    "write_blobs_read_logs": [
        {"blobs": "write"},
        {"logs": "read"},
    ],
[...]
            for svc, policy_id in policy_list:
                iam_role.add_managed_policy(managed_policies[svc][policy_id])
```

It avoids the string split, and the policy list cleanly maps onto the managed_policy's data structure.

Reply:
Agrh. I really hate forcing config writers to write extra { "", "" } where they can write in semi-plaintext ":".

I.e. yaml beats python :)

OK, if it will make you agree, I can do this. Not a huge difference

Reply:
Thanks, sorry for the trouble. :)

Reply:
No worries. It is fun.

Reply:
This way: see diff

Reply:
Looks good